toolchain/check/context.cpp

```diff
@@ -426,8 +426,9 @@ auto Context::LookupNameInExactScope(SemIRLoc loc, SemIR::NameId name_id,
                                      SemIR::NameScope& scope,
                                      bool is_being_declared)
     -> SemIR::ScopeLookupResult {
-  if (auto entry_id = is_being_declared ? scope.Lookup(name_id)
-                                        : scope.LookupOrPoison(name_id)) {
+  if (auto entry_id = is_being_declared || name_id == SemIR::NameId::SelfType
+                          ? scope.Lookup(name_id)
+                          : scope.LookupOrPoison(name_id)) {
```

Comment:
Can you point to a test where this fires? At trunk, I added here:

```
  CARBON_CHECK(!is_being_declared || (name_id == SemIR::NameId::Base ||
                                      name_id.AsIdentifierId().has_value()),
               "{0}", name_id);
```

And that seems to pass. But also, that could just be put into `LookupOrPoison` without changing this code if you wanted to just assert that the code was never used that way. What am I missing that you're special-casing `SelfType` here?

Reply:
How does this assert help checking this?

This is what I added, which fails:
```
  CARBON_CHECK(is_being_declared || name_id != SemIR::NameId::SelfType, "{0}",
               name_id);
```

I've encountered that when I tried to add poisoned names to the formatter and noticed that in a lot of cases, `Self` is being poisoned.
I don't think poisoning `Self` has an effect except costing more resources.

I wanted to clean this before trying to format poisoning names (which might be debateable regardless) so there we won't see a lot of file diffs because of that unnecessary poisoning.

Reply:
Sorry, maybe I'm being unclear here.

You're adding "|| name_id == SemIR::NameId::SelfType". I'm saying, I tried adding a CHECK that that condition was never true, and it seems to never be true. What am I missing?

Reply:
I've changed the logic of `LookupOrPoison()` to not poison non identifiers and improved the unit tests to properly demonstrate that.